### PR TITLE
first iteration to scope sdk styles to our components

### PR DESCRIPTION
**Pull request by npretto**
Most of the issues were caused by MetabaseProvider having css styles and passing it down to children, my solution was to remove all of them and moving them instead to the public component wrapper.

# Changes
- removes almost all css things applied to the provider
- moves them instead on the public component wrapper so that they don't leak into the host app
- adds to the public component wrapper some "reset" to prevent host app css to leak into our app

DEMO: https://www.loom.com/share/951197b0ef7e4112bac664f0d14865b9

Before:
- background color of the provider leaks in host app
- font family inside the provider is ours, not the host's
- font size also leaks from outside to inside (not shown)
image

After:

image

Code for the example:
```tsx
import { MetabaseProvider } from "@metabase/embedding-sdk-react";
import { Page } from "./Page";

const config = {
  metabaseInstanceUrl: "http://localhost:3000",
  jwtProviderUri: "http://localhost:8888/api/sso",
};

function App() {
  return (
    <div>
      <link rel="preconnect" href="https://fonts.googleapis.com" />
      <link
        rel="preconnect"
        href="https://fonts.gstatic.com"
        crossOrigin="true"
      />

      <link
        href="https://fonts.googleapis.com/css2?family=Sevillana&display=swap"
        rel="stylesheet"
      ></link>
      <p>outside of the provider</p>
      <style>
        {`
          body {
            font-family: Sevillana, serif;
            color: #FFAAAA;
            background-color: #000000;
            font-size: 30px;
             font-weight: bold;
          }
            p,h1,h2,h3,h4,h5,h6 {
             text-decoration: underline;
            }
        `}
      </style>
      <MetabaseProvider
        //
        theme={
          {
            // fontFamily: "Oswald, serif",
            // fontSize: "20px",
          }
        }
        config={config}
      >
        <hr />
        <p>inside of the provider, should be the same as above</p>
        <Page />
      </MetabaseProvider>
    </div>
  );
}

export default App;
```

```tsx
import {
  CreateQuestion,
  ModifyQuestion,
  StaticQuestion,
  useMetabaseAuthStatus,
} from "@metabase/embedding-sdk-react";

export const Page = () => {
  const userStatus = useMetabaseAuthStatus();
  return (
    <div>
      <h1>Page</h1>
      <pre>{JSON.stringify(userStatus)}</pre>
      <StaticQuestion questionId={15} />
      <StaticQuestion questionId={3} />

      <ModifyQuestion
        isSaveEnabled={true}
        questionId={1}
        onSave={(data) => console.log("Saved question", data)}
        onBeforeSave={async (data) => {
          console.log("Before save", data);
        }}
      />
    </div>
  );
};
```

### CODE REVIEW

File: enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx [hunk 1]
```diff
@@ -9,15 +9,40 @@ import { saveDomImageStyles } from "metabase/visualizations/lib/save-chart-image
  * even when rendered under a React portal.
  */
 export const PublicComponentStylesWrapper = styled.div`
+  // Try to reset as much as possible to avoid css leaking from host app to our components
+  all: initial;
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx:13]: These two line are a try to un-set things from the host app that we don't want.

File: enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx [hunk 2]
```diff
@@ -9,15 +9,40 @@ import { saveDomImageStyles } from "metabase/visualizations/lib/save-chart-image
...
+  font-style: normal;
+  -webkit-font-smoothing: antialiased;
+  -moz-osx-font-smoothing: grayscale;
+
   width: 100%;
   height: 100%;
 
   position: relative;
 
+  font-size: ${({ theme }) => theme.other.fontSize};
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx:37]: Otherwise font leaks from the host app, try to put font-size to 1 in the host app and you'll see some text disappear in our app too without this

File: enterprise/frontend/src/embedding-sdk/components/private/SdkGlobalStylesWrapper.tsx [hunk 3]
```diff
@@ -1,54 +0,0 @@
-import { css } from "@emotion/react";
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/components/private/SdkGlobalStylesWrapper.tsx:1]: Global styles were the reason of the issues so I removed this component and only added the few bits we needed (for fonts) to the provider

File: enterprise/frontend/src/embedding-sdk/components/private/SdkGlobalStylesWrapper.unit.spec.tsx [hunk 4]
```diff
@@ -1,61 +0,0 @@
-import { renderWithProviders, screen } from "__support__/ui";
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/components/private/SdkGlobalStylesWrapper.unit.spec.tsx:1]: I'll add the tests back if you approve of this "strategy", Ideally i'd put them on MetabaseProvider directly so that they're abstracted from lower level implementation details

File: enterprise/frontend/src/embedding-sdk/components/private/SdkThemeProvider.tsx [hunk 5]
```diff
@@ -33,13 +33,13 @@ export const SdkThemeProvider = ({ theme, children }: Props) => {
 
   return (
     <ThemeProvider theme={themeOverride}>
-      <SDKGlobalStyles />
+      <GlobalSdkCssVariables />
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/components/private/SdkThemeProvider.tsx:36]: It only contains css vars so I gave it a less scary name

File: enterprise/frontend/src/embedding-sdk/lib/theme/default-component-theme.ts [hunk 6]
```diff
@@ -138,5 +138,62 @@ export function getEmbeddingComponentOverrides(
         ...(theme?.popover?.zIndex && { zIndex: theme.popover.zIndex }),
       },
     },
+    // TODO: fix those, they're not all correct
+    ModalRoot: {
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/lib/theme/default-component-theme.ts:142]: I copy pasted the default props for all the components, but as you can probably tell some of them are broken (i'm passing target both as first level and nested in portalProps, I just wanted to have something i could test locally, I'll manually check which one are needed before we merge this

File: enterprise/frontend/src/embedding-sdk/lib/theme/default-component-theme.ts [hunk 7]
```diff
@@ -138,5 +138,62 @@ export function getEmbeddingComponentOverrides(
...
+      defaultProps: {
+        withinPortal: true,
+        // portalProps: { target: `#${EMBEDDING_SDK_ROOT_ELEMENT_ID}` },
+        target: `#${EMBEDDING_SDK_ROOT_ELEMENT_ID}`,
+      }, // satisfies Partial<ModalRootProps>,
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/lib/theme/default-component-theme.ts:147]: This is "fun", mantine doesn't typecheck default props for performance reason and suggests to use `sastifies xxx` if we want to typecheck default props.
Our version of prettier is old enough to fail if we use `satisfies`, i'll try and open a pr to update it

File: enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts [hunk 8]
```diff
@@ -21,7 +21,8 @@ import { SDK_TO_MAIN_APP_COLORS_MAPPING } from "./embedding-color-palette";
 const getFontFamily = (theme: MetabaseTheme) =>
   theme.fontFamily ?? DEFAULT_FONT;
 
-const SDK_BASE_FONT_SIZE = `${DEFAULT_SDK_FONT_SIZE / 16}em`;
+// const SDK_BASE_FONT_SIZE = `${DEFAULT_SDK_FONT_SIZE / 16}em`;
+const SDK_BASE_FONT_SIZE = `${DEFAULT_SDK_FONT_SIZE}px`;
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:25]: I don't think the default font size of the sdk should be in rem, otherwise it will be influenced by the font size of the host app.
Some people set the font size to 10px so that they can use 1.6em to mean 16px, I don't think it would work nice with our app.
I think we should use a fixed font size, and if the customer explicitly wants to change the font size then they can use the theme.

File: enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx [hunk 9]
```diff
@@ -9,15 +9,40 @@ import { saveDomImageStyles } from "metabase/visualizations/lib/save-chart-image
...
+  text-decoration: none;
+
+  // # Basic css reset
+  // We can't apply a global css reset as it would leak into the host app
+  // but we can't also apply our entire css reset scoped to this container,
+  // as it would be of higher specificity than some of our styles.
+  // We'll have to hand pick the css resets that we neeed
+
+  button {
+    border: 0;
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx:23]: These two lines fix the borders we get on buttons:
image

File: enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx [hunk 10]
```diff
@@ -76,20 +84,61 @@ export const MetabaseProviderInternal = ({
...
+    </EmotionCacheProvider>
   );
 };
 
 export const MetabaseProvider = memo(function MetabaseProvider(
   props: MetabaseProviderProps,
 ) {
-  return <MetabaseProviderInternal store={store} {...props} />;
+  return (
+    <Provider store={store}>
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx:103]: I did this so we could use useSelector directly in the internal provider, I'm realizing now that it's making unit tests to fail

File: enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx [hunk 11]
```diff
@@ -9,15 +9,38 @@ import { saveDomImageStyles } from "metabase/visualizations/lib/save-chart-image
  * even when rendered under a React portal.
  */
 export const PublicComponentStylesWrapper = styled.div`
+  // Try to reset as much as possible to avoid css leaking from host app to our components
+  all: initial;
+  text-decoration: none;
+
+  // # Basic css reset
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx:16]: I was very sad when I realized that we can't "simply" copy paste the css reset and have it scoped to our components 🙃

File: enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx [hunk 12]
```diff
@@ -9,15 +9,38 @@ import { saveDomImageStyles } from "metabase/visualizations/lib/save-chart-image
...
+  // as it would be of higher specificity than some of our styles.
+  // We'll have to hand pick the css resets that we neeed
+
+  button {
+    border: 0;
+    background-color: transparent;
+  }
+  // end of RESET
+
+  font-style: normal;
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx:28]: the css properties added below this are the ones i found on `rootStyle`, that was previously added at the MetabaseProvider level.
I initially had placed the entire rootStyle here but most of the properties were duplicated or not needed

File: enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx (hunk 1 above)
Comment (heypoom) [reply enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx:13]: woah, TIL that this property exists. I'll try to test with Shoppy to see if there are some unintended consequences of this

File: enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx (hunk 11 above)
Comment (heypoom) [reply enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx:16]: I see, is that because some selectors will leak outside of the PublicComponentStylesWrapper component?

File: enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx [hunk 13]
```diff
@@ -48,6 +55,7 @@ export const MetabaseProviderInternal = ({
   className,
 }: InternalMetabaseProviderProps): JSX.Element => {
   const { fontFamily = DEFAULT_FONT } = theme ?? {};
+  useInitData({ config });
```
Comment (heypoom) [inline enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx:58]: Is `useInitData` not already added to some other component?

File: enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts (hunk 8 above)
Comment (heypoom) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:25]: @npretto I recalled this is more of a **reference value** to [allow font size to scale in all viz components](https://github.com/metabase/metabase/pull/43264) though. It's not what determines the font size. If you use `px` here, then the viz elements' typography will no longer scale correctly.

**Review by heypoom — COMMENTED** [review]
Shared some initial feedback 🙏🏻

File: enterprise/frontend/src/embedding-sdk/lib/theme/default-component-theme.ts [hunk 14]
```diff
@@ -138,5 +138,62 @@ export function getEmbeddingComponentOverrides(
         ...(theme?.popover?.zIndex && { zIndex: theme.popover.zIndex }),
       },
     },
+    // TODO: fix those, they're not all correct
+    ModalRoot: {
+      defaultProps: {
+        withinPortal: true,
```
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/lib/theme/default-component-theme.ts:144]: I don't think this is going to work in all cases, there are some cases that a floating element needs to not be in the portal, and a case that it needs to be in the modal. This might break some of the default behavior for modal/popover/tooltip in some cases.

I'm just worried we're digging ourselves a hole with this. Since the combinations of floating elements are quite endless.

What I did for the main app is to inject styles directly into these floating element roots.

For example, this is for setting tooltip styles.
https://github.com/metabase/metabase/blob/7bd720439edd6254b13f1598a0f5bc689c73f9cb/frontend/src/metabase/public/components/EmbedFrame/EmbedFrame.module.css#L33

That's one more option in case you want to try it.

File: enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx (hunk 13 above)
Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx:58]: It was used here: https://github.com/metabase/metabase/pull/47764/files#diff-5c5c38f17eca5fe21b81ce25c487a3ab8f587a3200da195887158d7b9fe6397bL1
But we don't need that component anymore

File: enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx (hunk 2 above)
Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/components/private/PublicComponentStylesWrapper.tsx:35]: We set a default value on one of the many layers we have on the theme 🫠 
https://github.com/metabase/metabase/blob/91fad45f1a6fb712c8cbe73e42680b9275be1645/enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts#L46-L47

File: enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts (hunk 8 above)
Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:25]: Mmm let me check. Maybe we should put the font size in px in the container but use em inside so that the ems are relative to the px value 🤔.
The whole em/rem thing has always been confusing to me

Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:25]: @heypoom is this the expected behaviour?

https://github.com/user-attachments/assets/f58417fa-d9a1-4f55-b60d-570e957b7362

File: enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx [hunk 15]
```diff
@@ -47,7 +52,8 @@ export const MetabaseProviderInternal = ({
   store,
   className,
 }: InternalMetabaseProviderProps): JSX.Element => {
-  const { fontFamily = DEFAULT_FONT } = theme ?? {};
+  const { fontFamily } = theme ?? {};
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx:55]: I'm removing the default here, it's because otherwise we'll be saving the default font redux, and we'll never fallback to the one coming from the instance settings

File: enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx (hunk 10 above)
Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx:101]: (I fixed the tests)

File: enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx [hunk 16]
```diff
@@ -76,20 +82,33 @@ export const MetabaseProviderInternal = ({
...
+    <Provider store={store}>
+      <MetabaseProviderInternal store={store} {...props} />
+    </Provider>
+  );
 });
+
+/**
+ * This is the portal container used by popovers modals etc, it is wrapped with withPublicComponentWrapper
+ * so that it has our styles applied.
+ * Mantine components needs to have the defaultProps set to use `EMBEDDING_SDK_ROOT_ELEMENT_ID` as target for the portal
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx:110]:
```suggestion
 * Mantine components needs to have the defaultProps set to use `EMBEDDING_SDK_PORTAL_CONTAINER_ELEMENT_ID` as target for the portal
```

File: enterprise/frontend/src/embedding-sdk/config.ts [hunk 17]
```diff
@@ -1,5 +1,7 @@
 export const DEFAULT_FONT = "Lato";
 export const EMBEDDING_SDK_ROOT_ELEMENT_ID = "metabase-sdk-root";
+export const EMBEDDING_SDK_PORTAL_CONTAINER_ELEMENT_ID =
```
Comment (npretto) [inline enterprise/frontend/src/embedding-sdk/config.ts:3]: I added a new ID specific for the portal container for two reasons:
- it's technically not the root of the sdk anymore so it was a lie
- related to the above, the e2e tests were using that ID to scope down selectors, I decided not to change that so we won't have conflicts with the tests @WiNloSt has in his branch

File: enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts (hunk 8 above)
Comment (heypoom) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:21]: @npretto For the Bar Charts, yes that looks good. The trend chart's elements is also supposed to scale though if I remembered correctly, so that is a little weird. Let me check in production if trend charts scale properly in master.

File: enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts [hunk 18]
```diff
@@ -18,25 +18,23 @@ import {
...
+  font?: string,
 ): MantineThemeOverride {
   const components: MetabaseComponentTheme = merge(
     DEFAULT_EMBEDDED_COMPONENT_THEME,
     theme.components,
   );
 
   const override: MantineThemeOverride = {
-    fontFamily: getFontFamily(theme),
+    fontFamily: theme.fontFamily ?? font ?? DEFAULT_FONT,
```
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:37]: This is quite confusing why there are so many fonts. From my understanding:
- `theme` is what users passed to our SDK that might or might not include `font` so we need
- `font` this is from Metabase instance `font`, maybe it's better to name this `metabasteInstanceFont` or something
- `DEFAULT_FONT` then might not be needed since instance font should at least always be the fallback?

File: enterprise/frontend/src/embedding-sdk/components/private/SdkThemeProvider.tsx [hunk 19]
```diff
@@ -28,18 +29,18 @@ export const SdkThemeProvider = ({ theme, children }: Props) => {
     // This must be done before ThemeProvider calls getThemeOverrides.
     setGlobalEmbeddingColors(theme?.colors, appColors);
 
-    return theme && getEmbeddingThemeOverride(theme);
-  }, [appColors, theme]);
+    return theme && getEmbeddingThemeOverride(theme, font);
```
[the remark this answers is not in the document]
Comment (WiNloSt) [reply enterprise/frontend/src/embedding-sdk/components/private/SdkThemeProvider.tsx:32]: I've made another comment in `get-embedding-theme.ts`, I think we should rename this `instanceFont` or `metabaseInstanceFont` if the character limit isn't there.

This should help convey we either have font from `theme` (prop) or the instance font as a fallback.

```suggestion
    return theme && getEmbeddingThemeOverride(theme, instanceFont);
```

File: enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx (hunk 10 above)
Comment (WiNloSt) [reply enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx:101]: I read the first comment 2 times on different days, and I just got it now, I think this is needed because `useInitData` (which is just added to `MetabaseProviderInternal`) uses Redux under the hood, that's why.

File: enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx [hunk 20]
```diff
@@ -76,20 +82,33 @@ export const MetabaseProviderInternal = ({
...
 });
+
+/**
+ * This is the portal container used by popovers modals etc, it is wrapped with withPublicComponentWrapper
+ * so that it has our styles applied.
+ * Mantine components needs to have the defaultProps set to use `EMBEDDING_SDK_PORTAL_CONTAINER_ELEMENT_ID` as target for the portal
+ */
+const PortalContainer = withPublicComponentWrapper(() => (
+  <div id={EMBEDDING_SDK_PORTAL_CONTAINER_ELEMENT_ID}></div>
+));
```
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx:114]: This looks a bit overkill, well not by much, but hear me out.

`withPublicComponentWrapper` wraps whatever's inside with, obviously, a `PublicComponentWrapper`, and passes through whatever props from this component to the component inside, since we don't pass any prop here, I think using `PublicComponentWrapper` here might make the intent a bit clearer.

```suggestion
const PortalContainer = () => (
  <PublicComponentWrapper>
    <div id={EMBEDDING_SDK_PORTAL_CONTAINER_ELEMENT_ID}></div>
  </PublicComponentWrapper>
);
```

Another benefit is that I don't have to go inside this HoC util to figure out what it does.

File: enterprise/frontend/src/embedding-sdk/config.ts [hunk 21]
```diff
@@ -1,5 +1,7 @@
...
+  "metabase-sdk-portal-container";
```
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/config.ts:4]: I think you should still name this portal root or something. Since you're only rendering this portal container once per Metabase provider.

I think container has a broader meaning and there could be more than 1 containers like one container per popover for example, but root feels like there should be only a single root.

Another reason is that it makes this easier to understand even if a bit if you have both metabase root and portal root, you know they're roots for different purposes.

tl;dr it's just a naming thing, but I think `root` fits better.

File: frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx [hunk 22]
```diff
@@ -180,7 +180,7 @@ class TableInteractive extends Component {
 
     if (this.props.isEmbeddingSdk) {
       const rootElement = document.getElementById(
-        EMBEDDING_SDK_ROOT_ELEMENT_ID,
+        EMBEDDING_SDK_PORTAL_CONTAINER_ELEMENT_ID,
```
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx:183]: I guess the intent is to have everything styled, so we can measure the the actual size of this fake table. But we're rendering table here, and it doesn't have anything to do with portal.

Would keep using `EMBEDDING_SDK_ROOT_ELEMENT_ID` work in this case?

File: enterprise/frontend/src/embedding-sdk/config.ts (hunk 21 above)
Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/config.ts:4]: I'm fine with either name, but lest just remember that since we're using ids, I'm not exactly sure what will happen if have two providers on the same page.
It's probably ok if the theme passed to both is the same

Comment (WiNloSt) [reply enterprise/frontend/src/embedding-sdk/config.ts:4]: > but lest just remember that since we're using ids, I'm not exactly sure what will happen if have two providers on the same page.

When that happens, slapping some number prefix should help 😄 I'm not too concerned though.

File: frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx (hunk 22 above)
Comment (npretto) [reply frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx:183]: Yeah I think we need the styles and therefore what we currently call portal.
Technically if we're rendering stuff on that from another component it's a portal, although hand-made

Comment (WiNloSt) [reply frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx:183]: That's fair. Let's keep this.

File: enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts (hunk 18 above)
Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:37]: You're correct it's confusing and we have too many fonts 😆 , I think the fist bit (`theme.fontFamily`) is not needed actually, we store the font from the theme in redux: https://github.com/metabase/metabase/pull/47764/files#diff-022f68f2a8bede328179d2d89b7d3e3d41fd78a95fcfa46e3a4cf78ca5535a36R58-R62

For a moment I had set the name to `fontFromInstance` or something like that, but I realized that was a lie too as the font from the getFontSelector is either the one from the embed slice (which is the one from the theme) or the one from the instance.
I also think the only reason we need `DEFAULT_FONT` is to use while loading the settings, as the application-fonts has a default value of Lato.

Let me try and see if we can remove `theme.fontFamily` and if so I'll add a couple of comments that hopefully will make it easier to understand what's going on

Comment (WiNloSt) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:37]: >getFontSelector is either the one from the embed slice (which is the one from the theme) or the one from the instance.

The embed slice isn't set in the SDK though, since it's only called when we run Metabase frontend, in that sense `getFontSelector` on SDK is always an instance font (which might create even more confusion since it doesn't behave the same as in our Metabase FE).

Maybe creating a new selector entirely called `getInstanceFontSelector` might make things clearer. And we could use that in our SDK.

https://github.com/metabase/metabase/blob/54825a17312ee32702ea596fd20e3bdce3fee878/frontend/src/metabase/lib/embed.js#L45

Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:37]: It shouldn't be used as the `embed` slice is supposed to be for interactive, but we're using it for the sdk as well at the moment :(
Screenshot 2024-09-13 at 11 14 58

We set the "embed options" from the sdk here: https://github.com/metabase/metabase/blob/16b470da02b08410b59b690b9505b335e35f2a41/enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx#L58-L62
We should change that but I'd keep it as a next step otherwise we'll never stop fixing stuff

Comment (WiNloSt) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:37]: ok, this is another multi-layer code I really don't like since it's super hard to reason about :/ Thanks for pointing this out.

Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:37]: Yeah, I really want to get this merged, write tests that don't depend on any implementation, and then we can think about refactoring and removing layers without being scared of breaking stuff

File: enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx (hunk 10 above)
Comment (heypoom) [reply enterprise/frontend/src/embedding-sdk/components/public/MetabaseProvider.tsx:101]: Ah, so we lifted the `Provider` outside of the MetabaseProviderInternal for that -- that makes sense.

File: enterprise/frontend/src/embedding-sdk/lib/theme/default-component-theme.ts [hunk 23]
```diff
@@ -132,11 +137,31 @@ export function getEmbeddingComponentOverrides(
...
         },
 
         ...(theme?.popover?.zIndex && { zIndex: theme.popover.zIndex }),
       },
     },
+    ModalRoot: {
+      defaultProps: {
+        withinPortal: true,
+        target: `#${EMBEDDING_SDK_PORTAL_CONTAINER_ELEMENT_ID}`,
+      }, //  satisfies Partial<ModalRootProps>,
```
Comment (heypoom) [inline enterprise/frontend/src/embedding-sdk/lib/theme/default-component-theme.ts:150]: I think the `satisfies` PR have been merged, so you can use that now 👍🏻

File: enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts (hunk 18 above)
Comment (heypoom) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:37]: @npretto `theme.fontFamily` is just for Mantine's components such as `Text` and `Title`, so it should always be identical to the Redux store, and therefore does not need to be referenced here. Just referencing the font from Redux as you did with `font` should be 100% okay.

**Review by heypoom — COMMENTED** [review]
Looks good from my end! I'll test this with Shoppy and my demo app tonight and see if it works without unintended breaking changes.

File: enterprise/frontend/src/embedding-sdk/lib/theme/default-component-theme.ts (hunk 23 above)
Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/lib/theme/default-component-theme.ts:150]: We can, but mantine docs says they don't typecheck these by default because of performance, I'm 80% sure that the issue is typing all of the components and we're fine with just adding the `satisfies`s we need, but I'd still keep them commented just to play it safe for now

File: enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts (hunk 18 above)
Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts]: Does the code I just pushed look good?
I tested if mantine components use the instance font by default and they do, I'm honestly not sure which piece of code makes that work though 🙃

Comment (heypoom) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts]: @npretto that looks good! 🎉

Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts]: I tried to add `,sans-serif` to the default and I wanted to test it, but if i block the properties request, the page only renders the error components that for some reason don't have the correct fonts.
I feel like this is something that will take time to investigate and we'll have to add it to the "todo" for now

File: enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts (hunk 8 above)
Comment (npretto) [reply enterprise/frontend/src/embedding-sdk/lib/theme/get-embedding-theme.ts:21]: @heypoom it loooks like that font is expressed in `rem`, and it seems like if people mess with the font-size on the html tag it still breaks everything in the sdk, but i don't think we can do much about it

**Review by heypoom — APPROVED** [review]
Thanks for fixing these issues and addressing the feedback, this looks great!
